---
title: Data
---

import GoslingSchema from '../assets/gosling.schema.json';
import {TableWrapper} from './DocComponments/table'

Users can specify the data of each visualization (i.e., `track`) through a `track.data` property.
```javascript
{
  "tracks":[{
    "data": {...}, // specify the data used in this track
    "mark": "rect",
    "color": ...,
    ...
  }]
}
``` 

## Supported Data Formats

For the flexible data exploration, Gosling supports two different kinds of datasets:

1. **Plain Datasets** (No HiGlass Server): These datasets can be directly used in Gosling without requiring any data preprocessing, 
including [CSV](#csv-no-higlass-server), [JSON](#json-no-higlass-server), and [BigWig](#bigwig-no-higlass-server). 

2. **Pre-aggregated Datasets** (HiGlass Server): These datasets are preprocessed for the scalable data exploration and require a HiGlass server to access them in Gosling,
including [Vector](#vector-require-higlass-server), [Multivec](#multivec-require-higlass-server), and [BEDDE](#beddb-require-higlass-server).
To learn more about preprocessing your data and setting up the server, please visit the [HiGlass website](https://docs.higlass.io/).


### CSV (No HiGlass Server)

Any small enough tabular data files, such as tsv, csv, BED, BEDPE, and GFF, can be loaded using `"csv"` data specification.

```javascript
{
  "tracks": [
    {
      "data": {
        "url": "https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv",
        "type": "csv",
        "chromosomeField": "Chromosome",
        "genomicFields": ["chromStart", "chromEnd"]
      },
      ...,
  }]
}

```

<TableWrapper objName='CSVData'/>


### JSON (No HiGlass Server)

This format allows users to include data directly in the Gosling's JSON specification.

```javascript
{
  "tracks":[{
    "data": {
      "type": "json",
      "chromosomeField": "Chromosome",
      "genomicFields": [
          "chromStart",
          "chromEnd"
      ]
      "values": [
        {
          "Chromosome": "chr1",
          "chromStart": 0,
          "chromEnd": 2300000,
          "Name": "p36.33",
          "Stain": "gneg"
        },
        {
          "Chromosome": "chr1",
          "chromStart": 2300000,
          "chromEnd": 5300000,
          "Name": "p36.32",
          "Stain": "gpos25"
        }
        .....
        ]
    },
    ... // other configurations of this track
  }]
}
```

<TableWrapper objName='JSONData'/>


### BigWig (No HiGlass Server)

```javascript
{
  "tracks":[{
    "data": {
      "url": 'https://s3.amazonaws.com/gosling-lang.org/data/4DNFIMPI5A9N.bw',
      "type": "bigwig",
      "column": "position",
      "value": "peak"
    },
    ... // other configurations of this track
  }]
}
```

<TableWrapper objName='BIGWIGData'/>



### Vector (Require HiGlass Server)

One-dimensional quantitative values along genomic position (e.g., bigwig) can be converted into HiGlass' `"vector"` format data. Find out more about this format at [HiGlass Docs](https://docs.higlass.io/data_preparation.html#bigwig-files).

```javascript
{
  "tracks":[{
    "data": {
      "url": 'https://resgen.io/api/v1/tileset_info/?d=VLFaiSVjTjW6mkbjRjWREA',
      "type": "vector",
      "column": "position",
      "value": "peak"
    },
    ... // other configurations of this track
  }]
}
```

<TableWrapper objName='VectorData'/>

### Multivec (Require HiGlass Server)

Two-dimensional quantitative values, one axis for genomic coordinate and the other for different samples, can be converted into HiGlass' `"multivec"` data. For example, multiple BigWig files can be converted into a single multivec file. You can also convert sequence data (FASTA) into this format where rows will be different nucleotide bases (e.g., A, T, G, C) and quantitative values represent the frequency. Find out more about this format at [HiGlass Docs](https://docs.higlass.io/data_preparation.html#multivec-files).

```javascript
{
  "tracks":[{
    "data": {
        "url": "https://resgen.io/api/v1/tileset_info/?d=UvVPeLHuRDiYA3qwFlm7xQ",
        "type": "multivec",
        "row": "sample",
        "column": "position",
        "value": "peak",
        "categories": ["sample 1", "sample 2", "sample 3", "sample 4"]
    },
    ...// other configurations of this track
  }]
}
```

<TableWrapper objName='MultivecData'/>

 
### BEDDB (Require HiGlass Server)

Regular BED, or similar, files can be pre-aggregated for the scalable data exploration. Find our more about this format at [HiGlass Docs](https://docs.higlass.io/data_preparation.html#bed-files).

```javascript
{
  "tracks":[{
    "data": {
      "url": "https://higlass.io/api/v1/tileset_info/?d=OHJakQICQD6gTD7skx4EWA",
      "type": "beddb",
      "genomicFields": [
          {"index": 1, "name": "start"},
          {"index": 2, "name": "end"}
      ],
      "valueFields": [
          {"index": 5, "name": "strand", "type": "nominal"},
          {"index": 3, "name": "name", "type": "nominal"}
      ],
      "exonIntervalFields": [
          {"index": 12, "name": "start"},
          {"index": 13, "name": "end"}
      ]
    },
    ... // other configurations of this track
  }]
}
```

<TableWrapper objName='BEDDBData'/>



## Data Transform

<span>
    Gosling supports a diverse set of data transforms, including 
    {GoslingSchema["definitions"]['DataTransform'].anyOf.map((d, i)=>{
      let name = d.$ref.replace('#/definitions/', '')
      /** 
        seperate camel-case names. 
        e.g., "JSONParseTransform" => "JSON Parse Transform"  
      */
      name = name.replace(/([a-z])([A-Z])/g, '$1 $2').replace(/([A-Z])([A-Z][a-z])/g, '$1 $2')
        // replace white space ' ' to '-' to avoid the %20 in url
        return <span key={name}> <a href={`#${name.replace(/ /g, '-').toLowerCase()}`}>
            {name}
        </a> {i==GoslingSchema["definitions"]['DataTransform'].anyOf.length-1 ? '. ': ', '} </span>})
    }
</span>


```javascript
{
  "tracks":[{
    "data": ...,
    // a list of data transforms can be applied to the data
    "dataTransform": [
          { "type": "filter", "field": "type", "oneOf": ["gene"] },
          { "type": "filter", "field": "strand", "oneOf": ["+"], "not": true }
    ],
    "mark": "rect",
    ...,
  }]
}
```

### Filter Transform

Users can apply three types of filters: `oneOf`, `inRange`, `include`.
Each filter transform has the following properties:

**Properties of One Of Filter**
<TableWrapper objName='OneOfFilter' includeDescription={true}/>

**Properties of In Range Filter**
<TableWrapper objName='RangeFilter' includeDescription={true}/>

**Properties of Include Filter**
<TableWrapper objName='IncludeFilter' includeDescription={true}/>

### Str Concat Transform

<TableWrapper objName='StrConcatTransform' includeDescription={true}/>

### Str Replace Transform 

<TableWrapper objName='StrReplaceTransform' includeDescription={true}/>

### Log Transform

<TableWrapper objName='LogTransform' includeDescription={true}/>

### Displace Transform 

<TableWrapper objName='DisplaceTransform' includeDescription={true}/>

### Exon Split Transform 

<TableWrapper objName='ExonSplitTransform' includeDescription={true}/>

### Coverage Transform

<TableWrapper objName='CoverageTransform' includeDescription={true}/>

### JSON Parse Transform

<TableWrapper objName='JSONParseTransform' includeDescription={true}/>

Apart from these data transforms, users can also aggregate data values (min, max, bin, mean, and count). [Read more about data aggregation](#x)


#### Types

##### Type:Datum

<TableWrapper objName='Datum' />

##### Type: BoundingBox
<TableWrapper objName='DisplaceTransform-properties-boundingBox' />